### PR TITLE
Make Optimization Barrier in place

### PR DESCRIPTION
**Pull request by JackCaoG**
This will make `optimization barrier` less awkward. However I tested this implementation and see that it does not help the memory saving. 

Looking at the HLO graph. I think the issue is that
```
 %opt-barrier.9085 = f32[64,1024,14,14]{3,2,1,0} opt-barrier(f32[64,1024,14,14]{3,2,1,0} %maximum.9084)                                                                                                                                        
  %constant.9086 = f32[] constant(0)                                                                                                                                                                                                            
  %reduce.9092 = f32[] reduce(f32[64,1024,14,14]{3,2,1,0} %opt-barrier.9085, f32[] %constant.9086), dimensions={0,1,2,3}, to_apply=%AddComputation.9088 
```

optimization barrier HLO happened at the very end. It should be between forward and backward which will prevent xla compiler to merge forward and backward graph. I need to dig a bit more into this..

fyi @ronghanghu

### CODE REVIEW

**Comment by ronghanghu** [comment]
I'm thinking that a potential reason for the gradient checkpointing failure might be as follows: this API would prevent fusion between the variable in `optimization_barrier_`, but doesn't prevent fusion of the forward and the backward graph before this variable.

Take the following case as an example, which contains two parts, and gradient checkpointing is applied to `f` in Part 1:
```
import torch
import torch.utils.checkpoint
import torch_xla.core.xla_model as xm

def f(x):
	# some arbitrary function that might benefit from gradient checkpointing
    return torch.abs(torch.sin(x * x) * torch.cos(x * x))

##### Part 1 begin
a = torch.ones(10, device=xm.xla_device(), requires_grad=True)
b = torch.utils.checkpoint.checkpoint(f, a)  # checkpoint f
##### Part 1 end

loss = b.sum()
xm.optimization_barrier_([loss])

##### Part 2 begin
c = torch.ones(10, device=xm.xla_device()) + loss
d = torch.sin(c * c) + torch.cos(c * c)
e = d.sum()
##### Part 2 end

e.backward()
xm.mark_step()
```

I'm thinking that a potential failure here is that `xm.optimization_barrier_` only prevents fusion between the forward pass of Part 1 and the forward pass of Part 2, but *perhaps fails to prevent fusion between the forward and the backward pass of Part 1* and hence fails to activate gradient checkpointing on `f` in Part 1. Could that be the case here?

**Comment by JackCaoG** [comment]
I dump the ir graph of above program and see
```
IR {
.....
  %7 = f32[10]{0} aten::abs(%6), location=f@test_grad_simple.py:7, ROOT=1
  %8 = f32[] aten::sum(%7), location=<module>@test_grad_simple.py:14, dimensions=(0), keep_reduced_dimensions=0, dtype=6
  %9 = f32[] xla::optimization_barrier(%8), location=optimization_barrier_@xla_model.py:1037, ROOT=2
  %10 = f32[] prim::Constant(), location=<module>@test_grad_simple.py:18, value=1
  %11 = f32[] aten::mul(%9, %10), location=<module>@test_grad_simple.py:18
.....
  %20 = f32[10]{0} aten::mul(%14, %14), location=<module>@test_grad_simple.py:19
  %21 = f32[10]{0} aten::sin(%20), location=<module>@test_grad_simple.py:19
  %22 = f32[10]{0} aten::add(%21, %19), location=<module>@test_grad_simple.py:19, ROOT=4
  %23 = f32[] aten::sum(%22), location=<module>@test_grad_simple.py:20, dimensions=(0), keep_reduced_dimensions=0, dtype=6, ROOT=5
  %24 = f32[] prim::Constant(), location=backward@__init__.py:175, value=1
  %25 = f32[10]{0} aten::expand(%24), location=backward@__init__.py:175, size=(10)
  %26 = f32[10]{0} aten::mul(%1, %1), location=f@test_grad_simple.py:7
.....
  %77 = f32[10]{0} aten::add(%76, %66), location=backward@__init__.py:175
  %78 = f32[10]{0} aten::add(%77, %62), location=backward@__init__.py:175, ROOT=6
}

```

optimization barrier is inserted at the end of part 1. However if I also do 
```
xm.optimization_barrier_([e])
```
after part 2 ends. I can see the `optimization barrier`being correctly inserted between forward and backward
```
...
  %7 = f32[10]{0} aten::abs(%6), location=f@test_grad_simple.py:7, ROOT=1
  %8 = f32[] aten::sum(%7), location=<module>@test_grad_simple.py:14, dimensions=(0), keep_reduced_dimensions=0, dtype=6
  %9 = f32[] xla::optimization_barrier(%8), location=optimization_barrier_@xla_model.py:1037, ROOT=2
....
  %22 = f32[10]{0} aten::add(%21, %19), location=<module>@test_grad_simple.py:19, ROOT=4
  %23 = f32[] aten::sum(%22), location=<module>@test_grad_simple.py:20, dimensions=(0), keep_reduced_dimensions=0, dtype=6
  %24 = f32[] xla::optimization_barrier(%23), location=optimization_barrier_@xla_model.py:1037, ROOT=5
  %25 = f32[] prim::Constant(), location=backward@__init__.py:175, value=1
  %26 = f32[10]{0} aten::expand(%25), location=backward@__init__.py:175, size=(10)
  %27 = f32[10]{0} aten::mul(%1, %1), location=f@test_grad_simple.py:7
```

the later case should work.. let me go back to my tpu example..

**Comment by JackCaoG** [comment]
I manually verified this worked on TPU. For more detail, please refer to https://github.com/pytorch/xla/issues/3486#issuecomment-1099773576. I will merge this for now to unblock the user.
